src/measure_theory/decomposition/lebesgue.lean

```diff
@@ -0,0 +1,502 @@
+      { rw [ne.def, ennreal.add_eq_top, not_or_distrib],
+        exact ⟨ne_of_lt (measure_lt_top _ _), ne_of_lt (measure_lt_top _ _)⟩ },
+      { exact ne_of_lt (measure_lt_top _ _) },
+      { exact ne_of_lt (measure_lt_top _ _) },
+      { exact ne_of_lt (measure_lt_top _ _) },
+      { rw with_density_apply _ (hA.inter hE₁),
+        exact hξle (A ∩ E) (hA.inter hE₁) },
+      { apply_instance } },
+  -- from this, we can show `ξ + ε * E.indicator` is a function in `measurable_le` with
+  -- integral greater than `ξ`
+    have hξε : ξ + E.indicator (λ _, ε) ∈ measurable_le μ ν,
+    { refine ⟨measurable.add hξm (measurable.indicator measurable_const hE₁), λ A hA, _⟩,
+      have : ∫⁻ a in A, (ξ + E.indicator (λ _, ε)) a ∂μ =
+            ∫⁻ a in A ∩ E, ε + ξ a ∂μ + ∫⁻ a in A ∩ Eᶜ, ξ a ∂μ,
+      { rw [lintegral_add measurable_const hξm, add_assoc,
+            ← lintegral_union (hA.inter hE₁) (hA.inter (hE₁.compl))
+              (disjoint.mono (set.inter_subset_right _ _) (set.inter_subset_right _ _)
+              disjoint_compl_right), set.inter_union_compl],
+        simp_rw [pi.add_apply],
+        rw [lintegral_add hξm (measurable.indicator measurable_const hE₁), add_comm],
+        refine congr_fun (congr_arg has_add.add _) _,
+        rw [set_lintegral_const, lintegral_indicator _ hE₁, set_lintegral_const,
+            measure.restrict_apply hE₁, set.inter_comm] },
+      conv_rhs { rw ← set.inter_union_compl A E },
+      rw [this, measure_union _ (hA.inter hE₁) (hA.inter hE₁.compl)],
+      { exact add_le_add (hε₂ A hA)
+          (hξle (A ∩ Eᶜ) (hA.inter hE₁.compl)) },
+      { exact disjoint.mono (set.inter_subset_right _ _) (set.inter_subset_right _ _)
+          disjoint_compl_right } },
+      have : ∫⁻ a, ξ a + E.indicator (λ _, ε) a ∂μ ≤ Sup (measurable_le_eval μ ν) :=
+        le_Sup ⟨ξ + E.indicator (λ _, ε), hξε, rfl⟩,
+  -- but this contradicts the maximality of `∫⁻ x, ξ x ∂μ`
+      refine not_lt.2 this _,
+      rw [hξ₁, lintegral_add hξm (measurable.indicator (measurable_const) hE₁),
+          lintegral_indicator _ hE₁, set_lintegral_const],
+      refine ennreal.lt_add_right _ (ennreal.mul_pos.2 ⟨ennreal.coe_pos.2 hε₁, hE₂⟩),
+      have := measure_lt_top (μ.with_density ξ) set.univ,
+      rwa [with_density_apply _ measurable_set.univ, measure.restrict_univ] at this },
+  -- since `μ.with_density ξ ≤ ν`, it is clear that `ν = ν₁ + μ.with_density ξ`
+  { rw hν₁, ext1 A hA,
+    rw [measure.coe_add, pi.add_apply, measure.sub_apply hA hle,
+        add_comm, ennreal.add_sub_cancel_of_le (hle A hA)] },
+end
+
+/-- The Lebesgue decomposition is unique. -/
+theorem have_lebesgue_decomposition_of_finite_measure_unique
+  {μ ν : measure α} [finite_measure μ] [finite_measure ν]
+  (s : measure α) (f : α → ℝ≥0∞) (hs : s ⊥ₘ ν) (hadd : μ = s + ν.with_density f) :
+  s = μ.singular_part ν ∧ ν.with_density f = ν.with_density (μ.radon_nikodym_deriv ν) :=
```

Comment:
I find it very hard to work out exactly what is meant by "unique" in this theorem. Can you maybe make it clear that you mean that s and f are uniquely determined?

Personally I would prefer the equiv framing of the statement, but that might be my inner constructivist working.

Reply:
I don't think we can have a equiv since the radon nikodym derivative is only essentially unique, not unique

Reply:
You could have an equiv if you stored `radon_nikodym_deriv : quotient (setoid.ker μ.radon_nikodym_deriv)` but that obviously starts to become hard to work with